core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java

```diff
@@ -126,23 +201,783 @@ public Response getInfo(@QueryParam(QUERY_RAW_CONFIGURATION) final Boolean rawCo
+            relativeOffset = 0;
+          }
+          // If no param "end" presents, the offset is relative to the beginning; otherwise, it is
+          // relative to the end of the file.
+          if (requestEnd.equals("")) {
+            offset = relativeOffset;
+          } else {
+            offset = fileInfo.getLength() - relativeOffset;
+          }
+          if (offset < 0) {
+            offset = 0;
+          } else if (offset > fileInfo.getLength()) {
+            offset = fileInfo.getLength();
+          }
+          try {
+            AlluxioURI absolutePath = new AlluxioURI(currentFileInfo.getAbsolutePath());
+            FileSystem fs = FileSystem.Factory.get();
+            String fileData;
+            URIStatus status = fs.getStatus(absolutePath);
+            if (status.isCompleted()) {
+              OpenFilePOptions options =
+                  OpenFilePOptions.newBuilder().setReadType(ReadPType.NO_CACHE).build();
+              try (FileInStream is = fs.openFile(absolutePath, options)) {
+                int len = (int) Math.min(5 * Constants.KB, status.getLength() - offset);
+                byte[] data = new byte[len];
+                long skipped = is.skip(offset);
+                if (skipped < 0) {
+                  // nothing was skipped
+                  fileData = "Unable to traverse to offset; is file empty?";
+                } else if (skipped < offset) {
+                  // couldn't skip all the way to offset
+                  fileData = "Unable to traverse to offset; is offset larger than the file?";
+                } else {
+                  // read may not read up to len, so only convert what was read
+                  int read = is.read(data, 0, len);
+                  if (read < 0) {
+                    // stream couldn't read anything, skip went to EOF?
+                    fileData = "Unable to read file";
+                  } else {
+                    fileData = WebUtils.convertByteArrayToStringWithoutEscape(data, 0, read);
+                  }
+                }
+              }
+            } else {
+              fileData = "The requested file is not complete yet.";
+            }
+            List<UIFileBlockInfo> uiBlockInfo = new ArrayList<>();
+            for (FileBlockInfo fileBlockInfo : mFileSystemMaster
+                .getFileBlockInfoList(absolutePath)) {
+              uiBlockInfo.add(new UIFileBlockInfo(fileBlockInfo));
+            }
+            response.setFileBlocks(uiBlockInfo).setFileData(fileData)
+                .setHighestTierAlias(mBlockMaster.getGlobalStorageTierAssoc().getAlias(0));
+          } catch (AlluxioException e) {
+            throw new IOException(e);
+          }
+          response.setViewingOffset(offset);
+          return response;
+        }
+
+        if (currentPath.isRoot()) {
+          response.setPathInfos(new UIFileInfo[0]);
+        } else {
+          String[] splitPath = PathUtils.getPathComponents(currentPath.toString());
```

Comment:
There is a bug in this section of the code. We need to create a new variable, ie. `pathComponent = new AlluxioURI(AlluxioURI.SEPARATOR)` to build the path components. If we tack on to the current path, we will get incorrect paths (for example, for `/a/b/c`, we want `/a`, `/a/b`, the current implementation will return `/a/b/c/a` and `/a/b/c/a/b`)

Reply:
oh ok, will fix